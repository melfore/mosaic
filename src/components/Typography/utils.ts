type VariantDefaultsType = {
  [key: string]: {
    bottomSpacing: boolean;
    truncated: boolean;
  };
};

const VARIANT_DEFAULTS: VariantDefaultsType = {
  body1: {
    bottomSpacing: false,
    truncated: false,
  },
  caption: {
    bottomSpacing: false,
    truncated: false,
  },
  h5: {
    bottomSpacing: true,
    truncated: true,
  },
  h6: {
    bottomSpacing: true,
    truncated: true,
  },
  overline: {
    bottomSpacing: false,
    truncated: false,
  },
  subtitle1: {
    bottomSpacing: false,
    truncated: false,
  },
};

export const getBottomSpacing = (variant: string, bottomSpacing?: boolean): boolean =>
  bottomSpacing === undefined ? VARIANT_DEFAULTS[variant].bottomSpacing : bottomSpacing;

export const getTruncate = (variant: string, truncated?: boolean): boolean =>
  truncated === undefined ? VARIANT_DEFAULTS[variant].truncated : truncated;

export const VARIANT_COMPONENT_MAP = {
  body1: "p",
  caption: "span",
  h5: "h1",
  h6: "h2",
  overline: "h4",
  subtitle1: "h3",
};
