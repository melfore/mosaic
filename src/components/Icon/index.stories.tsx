import React from 'react';
import {  } from '@storybook/addon-actions';
import { select } from '@storybook/addon-knobs';
import { Icons, IconSize } from '../../types/Icon';
import { DOCS_PAGE_STRUCTURE, StoriesWrapper } from '../../utils/storybook';
import Icon from '.';

export default {
  title: 'Icon',
  component: Icon,
  parameters: {
    ...DOCS_PAGE_STRUCTURE,
  },
}

export const Canvas = () => (
  <Icon
    dataCy="icon"
    name={select('icon', Icons, Icons.add)}
    size={select('size', IconSize, IconSize.default)}
  />
);

export const Size = () => (
  <StoriesWrapper>
    <Icon
      dataCy="icon-send"
      name={Icons.send}
      size={IconSize.small}
    />
    <Icon
      dataCy="icon-send"
      name={Icons.send}
    />
    <Icon
      dataCy="icon-send"
      name={Icons.send}
      size={IconSize.large}
    />
  </StoriesWrapper>
);

const allIcons = Object
  .keys(Icons)
  .map((icon: string) => (
    <div className="icon-wrapper">
      <Icon
        dataCy={`icon-${icon}`}
        name={Icons[icon]}
      />
      <span>
        {icon}
      </span>
    </div>
  ));

export const AllIcons = () => (
  <StoriesWrapper>
    {allIcons}
  </StoriesWrapper>
);
