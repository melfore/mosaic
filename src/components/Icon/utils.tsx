import React from "react";
import MUIAddIcon from "@material-ui/icons/Add";
import MUIDeleteIcon from "@material-ui/icons/Delete";
import MUIEditIcon from "@material-ui/icons/Edit";
import MUIPlaylistAddCheckIcon from "@material-ui/icons/PlaylistAddCheck";
import MUISendIcon from "@material-ui/icons/Send";
import { Icons } from "../../types/Icon";

export const iconsCatalog = {
  [Icons.add]: <MUIAddIcon />,
  [Icons.delete]: <MUIDeleteIcon />,
  [Icons.edit]: <MUIEditIcon />,
  [Icons.playlistAddCheck]: <MUIPlaylistAddCheckIcon />,
  [Icons.send]: <MUISendIcon />,
};
