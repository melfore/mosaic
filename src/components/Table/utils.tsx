import React, { forwardRef, Fragment } from "react";
import { Action as MTActionType, Column as MTColumnType, Options as MTOptionsType } from "material-table";
import { Icons, IconSize } from "../../types/Icon";
import { TableActionScope, TableActionType, TableColumnType } from "../../types/Table";
import Button from "../Button";
import Icon from "../Icon";
import IconButton from "../IconButton";

// For default props refer to Options section here https://material-table.com/#/docs/all-props
export const DEFAULT_TABLE_OPTIONS: MTOptionsType = {
  debounceInterval: 500,
  draggable: false,
  emptyRowsWhenPaging: false,
};

export const actionAdapter = (action: TableActionType): MTActionType<object> => {
  const {
    callback,
    label,
    disabled = false,
    hidden = false,
    icon = undefined,
    scope = TableActionScope.default,
  } = action;
  return {
    disabled,
    hidden,
    icon: `${label}$${icon}`,
    isFreeAction: scope === TableActionScope.default,
    onClick: callback,
  };
};

export const actionComponentAdapter = (props: any) => {
  const { data, action } = props;
  const { disabled, hidden, icon, isFreeAction, onClick }: MTActionType<object> = action;
  if (hidden) {
    return null;
  }

  const labelAndIconName = icon as string;
  const label = labelAndIconName.split("$")[0];
  const iconName = labelAndIconName.split("$")[1] as Icons;
  return (
    <Fragment>
      {isFreeAction && (
        <Fragment>
          <span>&nbsp;&nbsp;&nbsp;&nbsp;</span>
          <Button
            disabled={disabled}
            icon={!iconName ? undefined : { name: iconName }}
            label={label}
            onClick={(event: any) => {
              onClick(event, data);
              event.preventDefault();
              event.stopPropagation();
            }}
          />
        </Fragment>
      )}
      {!isFreeAction && (
        <Fragment>
          <IconButton
            disabled={disabled}
            icon={iconName}
            onClick={(event: any) => {
              onClick(event, data);
              event.preventDefault();
              event.stopPropagation();
            }}
            size={IconSize.small}
          />
          <span>&nbsp;&nbsp;</span>
        </Fragment>
      )}
    </Fragment>
  );
};

export const columnAdapter = (column: TableColumnType): MTColumnType<object> => {
  const { label, path, render } = column;
  return {
    field: path,
    render,
    title: label,
  };
};

export const iconAdapter = (name: Icons, size = IconSize.default) =>
  forwardRef((props: any, ref: any) => {
    return <Icon forwarded={{ ...props, ref }} name={name} size={size} />;
  });
