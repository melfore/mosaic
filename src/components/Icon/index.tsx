import React, { FC } from "react";
import { IconType, IconSize } from "../../types/Icon";
import { iconsCatalog } from "./utils";

const Icon: FC<IconType> = ({ dataCy, forwarded = {}, name, size = IconSize.default }) => {
  const icon = iconsCatalog[name];
  return React.cloneElement(icon, { ...forwarded, "data-cy": dataCy, fontSize: size });
};

export default Icon;
