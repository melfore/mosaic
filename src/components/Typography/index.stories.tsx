import React from "react";
import {} from "@storybook/addon-actions";
import { text, select, boolean } from "@storybook/addon-knobs";
import { TypographyVariants } from "../../types/Typography";
import IntlProviderMock, { LocaleMock, MessageMock } from "../../utils/mocks/IntlProviderMock";
import { DOCS_PAGE_STRUCTURE, StoriesWrapper } from "../../utils/stories";
import Typography, { TypographyIntl } from ".";

export default {
  title: "Typography",
  component: Typography,
  parameters: {
    ...DOCS_PAGE_STRUCTURE,
  },
};

export const Canvas = () => (
  <Typography
    bottomSpacing={boolean("bottomSpacing", false)}
    label={text("label", "Typography example")}
    truncated={boolean("truncated", false)}
    variant={select("variant", TypographyVariants, TypographyVariants.body)}
  />
);

const allVariantsKeys = Object.keys(TypographyVariants);
const allVariantsValues = Object.values(TypographyVariants);
const allVariants: { [key: string]: string } = {};
allVariantsKeys.forEach((key, index) => (allVariants[key] = allVariantsValues[index]));
const getVariant = (variantKey: string) => allVariantsValues.find((vv) => vv === allVariants[variantKey]);
const allVariantsJsx = allVariantsKeys.map((variantKey) => (
  <div className="typography-wrapper">
    <Typography label={`${variantKey}`} variant={TypographyVariants.overline} />
    <Typography
      dataCy={`typography-${variantKey}`}
      label={`This is an example text`}
      variant={getVariant(variantKey)}
    />
  </div>
));

export const AllVariants = () => <StoriesWrapper>{allVariantsJsx}</StoriesWrapper>;

export const Truncated = () => (
  <StoriesWrapper>
    <div className="typography-wrapper">
      <div style={{ width: "50%" }}>
        <Typography label="Trucates with respect to parent width (resize window to make it happen)" truncated />
      </div>
      <div style={{ width: "200px" }}>
        <Typography label="Truncated all chars after 200px" truncated />
      </div>
    </div>
  </StoriesWrapper>
);

export const WithBottomSpacing = () => (
  <StoriesWrapper>
    <div className="typography-wrapper">
      <Typography label="Some text without bottom spacing" />
      <Typography bottomSpacing label="Some text with bottom spacing" />
      <Typography label="Other text" />
    </div>
  </StoriesWrapper>
);

export const WithIntl = () => (
  // IntlProviderMock simulates external IntlProvider context
  // The `TypographyIntl` exported version of `Typography` uses `labelId` prop to get a localized label.
  // When using `TypographyIntl` the prop `dataCy` can be omitted as it defaults to `labelId` value.
  <StoriesWrapper>
    <IntlProviderMock locale={select("locale", LocaleMock, LocaleMock.en)}>
      <TypographyIntl labelId={MessageMock.typography} />
    </IntlProviderMock>
  </StoriesWrapper>
);
