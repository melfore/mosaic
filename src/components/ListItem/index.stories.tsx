import React from "react";
import { action } from "@storybook/addon-actions";
import { text, boolean, select } from "@storybook/addon-knobs";
import { Icons } from "../../types/Icon";
import { TypographyVariants } from "../../types/Typography";
import { getDocsPageStructure, StoriesWrapper } from "../../utils/stories";
import ListItem from ".";

export default {
  title: "ListItem",
  component: ListItem,
  parameters: {
    ...getDocsPageStructure("ListItem", false),
  },
};

export const Canvas = () => (
  <ListItem
    dataCy="list-item"
    dense={boolean("dense", false)}
    icon={select("icon", Icons, undefined)}
    loading={boolean("loading", false)}
    onClick={action("On ListItem click")}
    selected={boolean("selected", false)}
    title={text("title", "ListItem Title")}
    titleVariant={select("titleVariant", TypographyVariants, undefined)}
  />
);

export const Basic = () => (
  <StoriesWrapper>
    <ListItem title="Basic List Item" />
  </StoriesWrapper>
);

export const Dense = () => (
  <StoriesWrapper>
    <ListItem title="Basic List Item" />
    <ListItem dense title="Basic List Item" />
  </StoriesWrapper>
);

export const Loading = () => (
  <StoriesWrapper>
    <ListItem loading title="Basic List Item" />
    <ListItem icon={Icons.add} loading title="Basic List Item" />
  </StoriesWrapper>
);

export const Selected = () => (
  <StoriesWrapper>
    <ListItem selected title="Basic List Item" />
  </StoriesWrapper>
);

export const WithIcon = () => (
  <StoriesWrapper>
    <ListItem icon={Icons.add} title="Basic List Item" />
  </StoriesWrapper>
);

export const WithTitleVariant = () => (
  <StoriesWrapper>
    <ListItem icon={Icons.add} title="Basic List Item" titleVariant={TypographyVariants.caption} />
  </StoriesWrapper>
);
