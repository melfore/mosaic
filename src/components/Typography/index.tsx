import React, { FC } from "react";
import MUITypography from "@material-ui/core/Typography";
import { BaseIntlType } from "../../types/Base";
import { TypographyType, TypographyVariants } from "../../types/Typography";
import withIntl from "../../utils/hocs/withIntl";
import { getBottomSpacing, getTruncate, VARIANT_COMPONENT_MAP } from "./utils";

/**
 * Typography component made on top of `@material-ui/core/Typography`.
 */
const Typography: FC<TypographyType> = ({
  bottomSpacing = undefined,
  dataCy = "typography",
  label,
  truncated = undefined,
  variant = TypographyVariants.body,
}) => {
  return (
    <MUITypography
      className={`data-cy-${dataCy}`}
      gutterBottom={getBottomSpacing(variant, bottomSpacing)}
      noWrap={getTruncate(variant, truncated)}
      variant={variant}
      variantMapping={VARIANT_COMPONENT_MAP}
    >
      {label}
    </MUITypography>
  );
};

export const TypographyIntl: FC<TypographyType & BaseIntlType> = withIntl(Typography);

export default Typography;
