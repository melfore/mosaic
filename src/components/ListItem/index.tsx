import React, { FC, memo, ReactElement, useMemo } from "react";
import {
  ListItem as MUIListItem,
  ListItemIcon as MUIListItemIcon,
  ListItemText as MUIListItemText,
} from "@material-ui/core";
import { Skeleton as MUISkeleton } from "@material-ui/lab";
import { ListItemType } from "../../types/ListItem";
import { TypographyVariants } from "../../types/Typography";
import Typography from "../Typography";
import Icon from "../Icon";

/**
 * ListItem component made on top of `@material-ui/core/ListItem`
 */
const ListItem: FC<ListItemType> = ({
  dense = true,
  icon = undefined,
  loading = false,
  onClick = undefined,
  selected = false,
  title,
  titleVariant = TypographyVariants.body,
}) => {
  const text = useMemo(
    () =>
      typeof title !== "string" ? (
        loading ? (
          <MUISkeleton />
        ) : (
          title
        )
      ) : (
        <Typography bottomSpacing={false} label={title} loading={loading} truncated variant={titleVariant} />
      ),
    [loading, title, titleVariant]
  );

  return (
    <MUIListItem
      dense={dense}
      onClick={(event) => {
        event.preventDefault();
        event.stopPropagation();
        if (loading) {
          return;
        }

        onClick && onClick();
      }}
      selected={!loading && selected}
      style={{ cursor: !loading && onClick ? "pointer" : "default" }}
    >
      {icon && (
        <MUIListItemIcon>
          <Icon loading={loading} name={icon} />
        </MUIListItemIcon>
      )}
      <MUIListItemText disableTypography>{text}</MUIListItemText>
    </MUIListItem>
  );
};

export default memo(ListItem);
