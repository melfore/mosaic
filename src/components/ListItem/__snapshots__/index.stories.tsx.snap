// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display/ListItem Children smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content"
  >
    Collapisble Children
  </div>
</div>
`;

exports[`Display/ListItem Clickable play-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: pointer;"
>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content"
  >
    John Doe
  </div>
</div>
`;

exports[`Display/ListItem Dense smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemText-root MuiListItemText-dense css-mui-classname"
       data-cy="list-item-content"
  >
    John Doe
  </div>
</div>
`;

exports[`Display/ListItem Icon smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemIcon-root css-mui-classname">
    <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mui-classname"
         focusable="false"
         aria-hidden="true"
         viewbox="0 0 24 24"
         data-testid="AccountCircleIcon"
         data-cy="list-item-icon"
    >
      <path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6m0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20">
      </path>
    </svg>
  </div>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content"
  >
    John Doe
  </div>
</div>
`;

exports[`Display/ListItem IconCustom smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemIcon-root css-mui-classname">
    <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mui-classname"
         focusable="false"
         aria-hidden="true"
         viewbox="0 0 24 24"
         data-testid="StyleIcon"
         data-cy="list-item-icon"
         style="height: 24px; width: 24px;"
    >
      <path d="m2.53 19.65 1.34.56v-9.03l-2.43 5.86c-.41 1.02.08 2.19 1.09 2.61m19.5-3.7L17.07 3.98c-.31-.75-1.04-1.21-1.81-1.23-.26 0-.53.04-.79.15L7.1 5.95c-.75.31-1.21 1.03-1.23 1.8-.01.27.04.54.15.8l4.96 11.97c.31.76 1.05 1.22 1.83 1.23.26 0 .52-.05.77-.15l7.36-3.05c1.02-.42 1.51-1.59 1.09-2.6M7.88 8.75c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1m-2 11c0 1.1.9 2 2 2h1.45l-3.45-8.34z">
      </path>
    </svg>
  </div>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content"
  >
    John Doe
  </div>
</div>
`;

exports[`Display/ListItem IconLoading smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemIcon-root css-mui-classname">
    <span class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-mui-classname"
          style="width: 24px; height: 24px;"
    >
    </span>
  </div>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content-loading"
  >
    <span class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-mui-classname">
    </span>
  </div>
</div>
`;

exports[`Display/ListItem Loading smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content-loading"
  >
    <span class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-mui-classname">
    </span>
  </div>
</div>
`;

exports[`Display/ListItem Primary smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content"
  >
    John Doe
  </div>
</div>
`;

exports[`Display/ListItem Selected smoke-test 1`] = `
<div class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-mui-classname"
     tabindex="0"
     role="button"
     data-cy="list-item"
     style="cursor: default;"
>
  <div class="MuiListItemText-root css-mui-classname"
       data-cy="list-item-content"
  >
    John Doe
  </div>
</div>
`;
